### PR TITLE
RectifiedAdam: Store 'total_steps' hyperparameter as float

### LINKED ISSUE #2361: Failing to restore RectifiedAdam optimizer from checkpoint
**System information**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Linux Ubuntu 18.04
- TensorFlow version and how it was installed (source or binary): source, nightly
- TensorFlow-Addons version and how it was installed (source or binary): source, nightly
- Python version: 3.6.7
- Is GPU used? (yes/no): yes

**Describe the bug**

When restoring the RectifiedAdam optimizer from a checkpoint, restoring fails with the following error message:
```ValueError: Tensor conversion requested dtype float32 for Tensor with dtype int32: <tf.Tensor: shape=(), dtype=int32, numpy=0>```

The reason is that RectifiedAdam stores `total_steps` as an integer hyper-parameter:
https://github.com/tensorflow/addons/blob/master/tensorflow_addons/optimizers/rectified_adam.py#L134

OptimizerV2 however creates variables by calling add_weight without a dtype in `_create_hypers`. In other words, it assumes that all hyper-parameters are floating point parameters.

Storing the `total_steps` parameter as float fixes the problem for me.

### PR DESCRIPTION
**Pull request by olesalscheider**
# Description

When restoring the RectifiedAdam optimizer from a checkpoint, restoring fails with the following error message:
```ValueError: Tensor conversion requested dtype float32 for Tensor with dtype int32: <tf.Tensor: shape=(), dtype=int32, numpy=0>```

The reason is that RectifiedAdam stores `total_steps` as an integer hyper-parameter:
https://github.com/tensorflow/addons/blob/master/tensorflow_addons/optimizers/rectified_adam.py#L134

OptimizerV2 however creates variables by calling `add_weight` without a dtype in `_create_hypers`. In other words, it assumes that all hyper-parameters are floating point parameters.

Fixes #2361

## Type of change

- [x] Bug fix
- [ ] New Tutorial
- [ ] Updated or additional documentation
- [ ] Additional Testing
- [ ] New Activation and the changes conform to the [activation contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/activations/README.md#contribution-guidelines)
- [ ] New Callback and the changes conform to the [callback contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/callbacks/README.md#contribution-guidelines)
- [ ] New Image addition and the changes conform to the [image op contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/image/README.md#contribution-guidelines)
- [ ] New Layer and the changes conform to the [layer contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/layers/README.md#contribution-guidelines)
- [ ] New Loss and the changes conform to the [loss contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/losses/README.md#contribution-guidelines)
- [ ] New Metric and the changes conform to the [metric contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/metrics/README.md#contribution-guidelines)
- [ ] New Optimizer and the changes conform to the [optimizer contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/optimizers/README.md#contribution-guidelines)
- [ ] New RNN Cell and the changes conform to the [rnn contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/rnn/README.md#contribution-guidelines)
- [ ] New Seq2seq addition and the changes conform to the [seq2seq contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/seq2seq/README.md#contribution-guidelines)
- [ ] New Text addition and the changes conform to the [text op contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/text/README.md#contribution-guidelines)

# Checklist:

- [x] I've properly [formatted my code according to the guidelines](https://github.com/tensorflow/addons/blob/master/CONTRIBUTING.md#coding-style)
    - [x] By running Black + Flake8
    - [ ] By running pre-commit hooks
- [x] This PR addresses an already submitted issue for TensorFlow Addons
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] This PR contains modifications to C++ custom-ops

# How Has This Been Tested?

I have trained a neural network with this optimizer and resumed training in between. Without this change, restoring fails with the error message from above. With this change, it works as intended.

### CODE REVIEW

**Comment by olesalscheider** [comment]
The tests fail because `total_steps` is int in the constructor. I can fix that, but I am really not sure if this is the right way to handle this, or if this is unintended behavior from tensorflow.

**Review by WindQAQ — CHANGES_REQUESTED** [review]
Thanks! Can you add a test that failed before the fix?

File: tensorflow_addons/optimizers/rectified_adam.py [hunk 1]
```diff
@@ -175,7 +175,7 @@ def _resource_apply_dense(self, grad, var):
         beta_2_power = tf.pow(beta_2_t, local_step)
 
         if self._initial_total_steps > 0:
-            total_steps = self._get_hyper("total_steps", var_dtype)
+            total_steps = int(self._get_hyper("total_steps", var_dtype))
```
Comment (WindQAQ) [inline tensorflow_addons/optimizers/rectified_adam.py:178]: Why it needs to be casted to int? I think following computation all in `var_dtype`.

File: tensorflow_addons/optimizers/rectified_adam.py [hunk 2]
```diff
@@ -131,7 +131,7 @@ def __init__(
         self._set_hyper("decay", self._initial_decay)
         self._set_hyper("weight_decay", weight_decay)
         self._set_hyper("sma_threshold", sma_threshold)
-        self._set_hyper("total_steps", int(total_steps))
+        self._set_hyper("total_steps", float(total_steps))
```
Comment (WindQAQ) [inline tensorflow_addons/optimizers/rectified_adam.py:134]: If the weights are added in floating point, do we really need `float(xxx)` here?

**Comment by bhack** [comment]
It is a little bit confusing as a float.

**Comment by olesalscheider** [comment]
> It is a little bit confusing as a float.

That's what I thought. But if the constructor takes an int but stores it as a float internally, the tests fail because the optimizer can't be restored.

**Comment by bhack** [comment]
Mhh.. I've not run this but can we use a `_get_hyper` with the `dtype` before the set_hyper?

File: tensorflow_addons/optimizers/rectified_adam.py [hunk 3]
```diff
@@ -78,7 +78,7 @@ def __init__(
         weight_decay: Union[FloatTensorLike, Callable, Dict] = 0.0,
         amsgrad: bool = False,
         sma_threshold: FloatTensorLike = 5.0,
-        total_steps: int = 0,
+        total_steps: FloatTensorLike = 0.0,
```
Comment (WindQAQ) [inline tensorflow_addons/optimizers/rectified_adam.py:81]: How about `TensorLike = 0`?

File: tensorflow_addons/optimizers/tests/rectified_adam_test.py [hunk 4]
```diff
@@ -209,3 +210,26 @@ def test_scheduler_serialization():
         "class_name": "InverseTimeDecay",
         "config": wd_scheduler.get_config(),
     }
+
+
+def test_checkpoint_serialization():
```
Comment (WindQAQ) [inline tensorflow_addons/optimizers/tests/rectified_adam_test.py:215]: We get pytest fixture to access tmpdir. Try

https://github.com/tensorflow/addons/blob/9dd2389124a52e6cc298eba7d16c6732ec484d30/tensorflow_addons/layers/tests/spatial_pyramid_pooling_test.py#L67-L80

File: tensorflow_addons/optimizers/rectified_adam.py (hunk 3 above)
Comment (bhack) [reply tensorflow_addons/optimizers/rectified_adam.py:81]: If `_get_hyper` with dytpe doesn't work It could be a good compromise to relax the type hint

**Review by bhack — CHANGES_REQUESTED** [review]

**Comment by bhack** [comment]
@googlebot I consent

**Review by bhack — APPROVED** [review]

**Review by WindQAQ — APPROVED** [review]

File: tensorflow_addons/optimizers/rectified_adam.py [hunk 5]
```diff
@@ -78,7 +79,7 @@ def __init__(
         weight_decay: Union[FloatTensorLike, Callable, Dict] = 0.0,
         amsgrad: bool = False,
         sma_threshold: FloatTensorLike = 5.0,
-        total_steps: int = 0,
+        total_steps: TensorLike = 0.0,
```
[the remark this answers is not in the document]
Comment (ykpercy) [reply tensorflow_addons/optimizers/rectified_adam.py]: so you mean total_steps could input with tf.cast(100, tf.float32).
however, the code still is total_steps: int = 0. input dtype limit with int

**Pull request merged**